In [ ]:
!python -m pip install -U python-woc pandas matplotlib
# please clear the output of this cell in your notebook before checking it in

In [2]:
import sys
print(sys.executable)

/home/fancy/venvs/woc312/bin/python


In [1]:
from woc.remote import WocMapsRemote
from tqdm import tqdm
import pandas as pd
# creates the client
woc = WocMapsRemote( base_url="https://worldofcode.org/api/")
# if you got an API key
# woc = WocMapsRemote( base_url="https://worldofcode.org/api/", api_key="woc-XXXXXX-YYYYYY" )



In [18]:
# Now for each of the ten projects you were assigned get commits, e.g.
projects = ['tyiannak/pyAudioAnalysis','FZJ-INM1-BDA/siibra-explorer','spacetelescope/jwql', 'smhyatt/deforest-watch-with-BFAST-in-CUDA', 'eurostat/statistics-coded', 'tnagler/VineCopula', 'deeptime-ml/deeptime', 'aiidateam/aiida-tutorials', 'lepy/phuzzy', 'dpiparo/vdt']
list_df_commits = []
for prj in projects:
  # Convert GitHub repository names to WoC V2412 project names
  prj = prj.lower().replace('/','_',2)
  commits = woc.get_values('p2c', prj)
  df = pd.DataFrame(commits, columns=["sha1"])
  df['project'] = prj
  list_df_commits.append(df)
df = pd.concat(list_df_commits)
#perhaps save the list (if no errors) so you do not need to retrieve them again
df.to_csv('df_commits.csv')
df

,sha1,project
0,000a5c520d3dcf9cb1847f17092cb178032013f8,tyiannak_pyaudioanalysis
1,00204ff33b60af2b6da7b3f0ff82865fdcb971d1,tyiannak_pyaudioanalysis
2,0030b718e92eb6b4798f7fe0bef8b9757aad54e3,tyiannak_pyaudioanalysis
3,0046c875f107159789cafd14cf07d04f3ac195c4,tyiannak_pyaudioanalysis
4,0082170c5623bbcb7aa4cf5be46d82d67bc23943,tyiannak_pyaudioanalysis
...,...,...
310,fdb34ff87169659f67cd5c869b3dc21eb19767d1,dpiparo_vdt
311,fddf8aebb3269db1d8a3109eaaa3b1c88de382d7,dpiparo_vdt
312,fdea1cfcea097a09ddc362ab1ff21b10899fbad9,dpiparo_vdt
313,fe4efdcdaa8663193132cca1120ec55b12e4702e,dpiparo_vdt


In [19]:
# If the number of commits its not very large, you can try to get them all at the same time,
# The max batch size is 10
import time
# let us first split df['sha1'] in chunks
chunks = [df['sha1'][x:x+10] for x in range(0, len(df), 10)]
commit_data = []

for chunk in tqdm(chunks): # iterate over the commits
  # res, err = woc.show_content_many('commit',chunk.to_list())

  # commit.tch returns the same data but faster
  res, err = woc.get_values_many('commit.tch',chunk.to_list())
  res = {k: v[0] for k, v in res.items()}  # this conversion is necessary because of the internal implementation

  # to walk around the rate limit
  time.sleep(1)

  if err: # check for errors
    print('Got Errors', err)

  for commit_sha, commit in res.items():
    # flatten commit objects
    commit_data.append({
          'commit': commit_sha,
          'tree': commit[0],
          'parent': list(commit[1]),
          'author': commit[2][0],
          'author_time': int(commit[2][1]),
          'author_tz': commit[2][2],
          'committer': commit[3][0],
          'committer_time': int(commit[3][1]),
          'committer_tz': commit[3][2],
          'message': commit[4],
    })

df_commit_data = pd.DataFrame(commit_data)

df_commit_data = df_commit_data.merge(df, left_on='commit', right_on='sha1')
df_commit_data.to_csv('df_commit_data.csv', index=False)
df_commit_data.head(2)

 29%|██▉       | 643/2184 [11:15<26:56,  1.05s/it]

Got Errors {'11357dd6767ea3eea1136e98333fdf75b74fb2de': 'Key 11357dd6767ea3eea1136e98333fdf75b74fb2de not found in /da5_fast/All.sha1c/commit_17.tch'}


 30%|███       | 661/2184 [11:33<26:37,  1.05s/it]

Got Errors {'16b9cbeae06c0000ade73ced2291d67d9e61203c': 'Key 16b9cbeae06c0000ade73ced2291d67d9e61203c not found in /da5_fast/All.sha1c/commit_22.tch'}


 33%|███▎      | 724/2184 [12:40<25:30,  1.05s/it]

Got Errors {'2ab86ea220275aaff5acc7b45c5f15e06f4dade5': 'Key 2ab86ea220275aaff5acc7b45c5f15e06f4dade5 not found in /da5_fast/All.sha1c/commit_42.tch'}


 34%|███▍      | 748/2184 [13:05<25:08,  1.05s/it]

Got Errors {'304dbedd9d835f9672ac50e16621863077080e0e': 'Key 304dbedd9d835f9672ac50e16621863077080e0e not found in /da5_fast/All.sha1c/commit_48.tch'}


 36%|███▌      | 784/2184 [13:42<24:24,  1.05s/it]

Got Errors {'3b2543a2cf7041b28a237f95e511d1488920671c': 'Key 3b2543a2cf7041b28a237f95e511d1488920671c not found in /da5_fast/All.sha1c/commit_59.tch'}


 36%|███▌      | 786/2184 [13:45<24:32,  1.05s/it]

Got Errors {'3b9f663cd6406d4b2a0fa3dab94598a34f879b38': 'Key 3b9f663cd6406d4b2a0fa3dab94598a34f879b38 not found in /da5_fast/All.sha1c/commit_59.tch'}


 38%|███▊      | 834/2184 [14:35<23:46,  1.06s/it]

Got Errors {'4ad71052fad349591aa68d7ac879440c0d590abb': 'Key 4ad71052fad349591aa68d7ac879440c0d590abb not found in /da5_fast/All.sha1c/commit_74.tch'}


 38%|███▊      | 840/2184 [14:41<23:31,  1.05s/it]

Got Errors {'4c6b9dd9a49dfe009723441304ac87ccf4b6d359': 'Key 4c6b9dd9a49dfe009723441304ac87ccf4b6d359 not found in /da5_fast/All.sha1c/commit_76.tch'}


 40%|████      | 877/2184 [15:21<23:11,  1.06s/it]

Got Errors {'572b48c888d3f592e0d0f1e5a949b462e2b27929': 'Key 572b48c888d3f592e0d0f1e5a949b462e2b27929 not found in /da5_fast/All.sha1c/commit_87.tch'}


 44%|████▍     | 964/2184 [16:53<21:39,  1.07s/it]

Got Errors {'70389676607c079b065c263f115a61f78f6a8140': 'Key 70389676607c079b065c263f115a61f78f6a8140 not found in /da5_fast/All.sha1c/commit_112.tch'}


 45%|████▍     | 978/2184 [17:08<21:26,  1.07s/it]

Got Errors {'74f59000614d3d54a037465f1dd094b13dfbeed5': 'Key 74f59000614d3d54a037465f1dd094b13dfbeed5 not found in /da5_fast/All.sha1c/commit_116.tch'}


 45%|████▌     | 987/2184 [17:18<21:15,  1.07s/it]

Got Errors {'77557dc92db9277d58364884a4c224857f41b937': 'Key 77557dc92db9277d58364884a4c224857f41b937 not found in /da5_fast/All.sha1c/commit_119.tch'}


 47%|████▋     | 1026/2184 [18:00<20:29,  1.06s/it]

Got Errors {'8344a3527f7b7725fe9a619f69215f014d12dae1': 'Key 8344a3527f7b7725fe9a619f69215f014d12dae1 not found in /da5_fast/All.sha1c/commit_3.tch'}


 51%|█████     | 1105/2184 [19:24<19:05,  1.06s/it]

Got Errors {'9a9c0ea9bdaab9ed57d77f0b478802d67f2b7bdf': 'Key 9a9c0ea9bdaab9ed57d77f0b478802d67f2b7bdf not found in /da5_fast/All.sha1c/commit_26.tch'}


 54%|█████▍    | 1176/2184 [20:39<17:58,  1.07s/it]

Got Errors {'b14b1252b7d486c625272ffe1f5fea91a00ec39b': 'Key b14b1252b7d486c625272ffe1f5fea91a00ec39b not found in /da5_fast/All.sha1c/commit_49.tch'}


 55%|█████▌    | 1202/2184 [21:07<17:23,  1.06s/it]

Got Errors {'b87db8b8441adeca5343fb7f17220f423979fef5': 'Key b87db8b8441adeca5343fb7f17220f423979fef5 not found in /da5_fast/All.sha1c/commit_56.tch'}


 56%|█████▌    | 1215/2184 [21:21<17:10,  1.06s/it]

Got Errors {'bc9b587938e32ffc0c52215affb55f7b4adf3871': 'Key bc9b587938e32ffc0c52215affb55f7b4adf3871 not found in /da5_fast/All.sha1c/commit_60.tch'}


 56%|█████▌    | 1218/2184 [21:24<17:10,  1.07s/it]

Got Errors {'bd9c1a322772c171052d6018abb8a407afec53f3': 'Key bd9c1a322772c171052d6018abb8a407afec53f3 not found in /da5_fast/All.sha1c/commit_61.tch'}


 58%|█████▊    | 1271/2184 [22:20<16:07,  1.06s/it]

Got Errors {'ce151337f58b6420625461402f9ce60ba13db29f': 'Key ce151337f58b6420625461402f9ce60ba13db29f not found in /da5_fast/All.sha1c/commit_78.tch'}


 58%|█████▊    | 1274/2184 [22:23<16:02,  1.06s/it]

Got Errors {'cf36c10fecc65a974acd45b1572f339c0c3b5a7f': 'Key cf36c10fecc65a974acd45b1572f339c0c3b5a7f not found in /da5_fast/All.sha1c/commit_79.tch'}


 61%|██████▏   | 1341/2184 [23:35<14:58,  1.07s/it]

Got Errors {'e370012e459cd0693d9d85488849ade7785def6f': 'Key e370012e459cd0693d9d85488849ade7785def6f not found in /da5_fast/All.sha1c/commit_99.tch'}


 61%|██████▏   | 1342/2184 [23:36<14:59,  1.07s/it]

Got Errors {'e41d53a9a025d22c1c97644cab3a44f051acfe09': 'Key e41d53a9a025d22c1c97644cab3a44f051acfe09 not found in /da5_fast/All.sha1c/commit_100.tch'}


 63%|██████▎   | 1371/2184 [24:06<14:22,  1.06s/it]

Got Errors {'eb6dbaebb9bd493fa4986c92103d9c456dd132e9': 'Key eb6dbaebb9bd493fa4986c92103d9c456dd132e9 not found in /da5_fast/All.sha1c/commit_107.tch'}


 63%|██████▎   | 1386/2184 [24:22<14:07,  1.06s/it]

Got Errors {'f039d2474dd86c3a61c9040685bfbc15d4fbfb21': 'Key f039d2474dd86c3a61c9040685bfbc15d4fbfb21 not found in /da5_fast/All.sha1c/commit_112.tch'}


 65%|██████▍   | 1412/2184 [24:50<13:40,  1.06s/it]

Got Errors {'f743f67de9dc28616372a61c601c68d12468f653': 'Key f743f67de9dc28616372a61c601c68d12468f653 not found in /da5_fast/All.sha1c/commit_119.tch'}


 97%|█████████▋| 2120/2184 [37:24<01:08,  1.07s/it]

Got Errors {'4ded14204aaa26bea4e342ff18babbfe041a7a7c': 'Key 4ded14204aaa26bea4e342ff18babbfe041a7a7c not found in /da5_fast/All.sha1c/commit_77.tch', '4e2432c8290bd0cb5279c0abb220feab3db7eca7': 'Key 4e2432c8290bd0cb5279c0abb220feab3db7eca7 not found in /da5_fast/All.sha1c/commit_78.tch'}


 97%|█████████▋| 2129/2184 [37:34<00:58,  1.07s/it]

Got Errors {'81d6f0036687f170609fe29091d51c5629920e14': 'Key 81d6f0036687f170609fe29091d51c5629920e14 not found in /da5_fast/All.sha1c/commit_1.tch'}


 98%|█████████▊| 2146/2184 [37:52<00:40,  1.07s/it]

Got Errors {'df448cd56c406f69df49c41f60a298a9553874b3': 'Key df448cd56c406f69df49c41f60a298a9553874b3 not found in /da5_fast/All.sha1c/commit_95.tch'}


100%|██████████| 2184/2184 [38:32<00:00,  1.06s/it]


,commit,tree,parent,author,author_time,author_tz,committer,committer_time,committer_tz,message,sha1,project
0,000a5c520d3dcf9cb1847f17092cb178032013f8,8c428776f1a407f5891c40140b0895410e75c0a6,[0dfe4059ec3675d50f6ea46156170491bedc7d13],Theodoros Giannakopoulos <tyiannak@gmail.com>,1411566537,+0300,Theodoros Giannakopoulos <tyiannak@gmail.com>,1411566537,+0300,added arff output in features\n,000a5c520d3dcf9cb1847f17092cb178032013f8,tyiannak_pyaudioanalysis
1,00204ff33b60af2b6da7b3f0ff82865fdcb971d1,4d2443e49d96c7e2dc5956493a85bb56e16cb166,[abd33bbe25ef3d5869da6e0f717e45163660e0c6],Theodoros Giannakopoulos <tyiannak@gmail.com>,1432738598,+0300,Theodoros Giannakopoulos <tyiannak@gmail.com>,1432738598,+0300,movie added\n,00204ff33b60af2b6da7b3f0ff82865fdcb971d1,tyiannak_pyaudioanalysis


In [20]:
# what does the first record looks like?
# commit  parents
v = commit_data[0]
# commit tree, [parent{s}], [author, unixtime, tz ], [ committer, unixtime, tz ], Commit_message ]
print(v)

{'commit': '000a5c520d3dcf9cb1847f17092cb178032013f8', 'tree': '8c428776f1a407f5891c40140b0895410e75c0a6', 'parent': ['0dfe4059ec3675d50f6ea46156170491bedc7d13'], 'author': 'Theodoros Giannakopoulos <tyiannak@gmail.com>', 'author_time': 1411566537, 'author_tz': '+0300', 'committer': 'Theodoros Giannakopoulos <tyiannak@gmail.com>', 'committer_time': 1411566537, 'committer_tz': '+0300', 'message': 'added arff output in features\n'}


In [21]:
# Now that the data is retrieved, save it and commit to your GH fork
# First, we need to flatten info in order to export as csv
# our dataframe will have columns project,'commit, author, time, message'
dfinf = pd.DataFrame(columns=['project', 'commit', 'author', 'time', 'message'])
for k in commit_data:
  row = pd.Series({'project':prj, 'commit': k['commit'], 'author': k['author'], 'time': k['author_time'], 'message':k['message']})
  dfinf = pd.concat([dfinf, row.to_frame().T ], ignore_index=True)


In [22]:
# check if it has the right content
dfinf.head(1)

,project,commit,author,time,message
0,dpiparo_vdt,000a5c520d3dcf9cb1847f17092cb178032013f8,Theodoros Giannakopoulos <tyiannak@gmail.com>,1411566537,added arff output in features\n


In [23]:
#mode a means append, so you have all your projects in the same file
yournetid='hnaicker'
dfinf.to_csv(yournetid+'_project_summary.csv', index=False,sep=';', mode='a', header=False)

In [ ]:
#Code to get max and min times
df_commit_data['dt'] = pd.to_datetime(df_commit_data['author_time'], unit='s', utc=True)

time_range = df_commit_data.groupby('project')['dt'].agg(
    min_time='min',
    max_time='max',
).reset_index()

time_range

,project,min_time,max_time
0,aiidateam_aiida-tutorials,2018-05-28 16:53:57+00:00,2025-10-02 11:28:57+00:00
1,deeptime-ml_deeptime,2015-04-11 10:31:18+00:00,2025-04-14 02:31:18+00:00
2,dpiparo_vdt,2012-08-23 07:16:32+00:00,2025-05-03 22:18:58+00:00
3,eurostat_statistics-coded,2019-01-11 17:03:26+00:00,2025-03-20 13:04:12+00:00
4,fzj-inm1-bda_siibra-explorer,2017-07-24 08:51:16+00:00,2025-10-27 13:07:07+00:00
5,lepy_phuzzy,2017-06-28 07:38:13+00:00,2025-12-31 11:16:57+00:00
6,smhyatt_deforest-watch-with-bfast-in-cuda,2019-09-02 15:10:41+00:00,2025-03-18 17:41:59+00:00
7,spacetelescope_jwql,2017-11-13 16:39:24+00:00,2025-11-01 06:00:04+00:00
8,tnagler_vinecopula,2015-09-23 19:33:27+00:00,2025-08-07 13:35:26+00:00
9,tyiannak_pyaudioanalysis,2014-08-27 12:44:23+00:00,2025-08-18 14:51:43+00:00


# Make sure you check in to your fork not just the notebook but also the csv files!!!

# Don't forget to add requested data from github and this notebook
### For each of the 10 projects go to their github repo and get the number of stars, number of forks, and the last commit date
### Report (in your notebook) the number of commits, the number of authors, and max and min time for each project based on WoC commits and also add the info you obtained from github

FORMAT:

Git link

[number of stars, number of forks, last commit date]

[number of commits, number of authors, max time, min time]


https://github.com/tyiannak/pyAudioAnalysis :

[6.3k, 1.2k, 8/04/2025]

[782, 22, 2025-08-18 14:51:43+00:00, 2014-08-27 12:44:23+00:00]


https://github.com/FZJ-INM1-BDA/siibra-explorer :

[30, 6, 07/26/2026]

[3728, 6, 2025-10-27 13:07:07+00:00, 2017-07-24 08:51:16+00:00]

https://github.com/spacetelescope/jwql :

[68, 47, 08/15/2026]

[7880, 26, 2025-11-01 06:00:04+00:00, 2017-11-13 16:39:24+00:00]

https://github.com/smhyatt/deforest-watch-with-BFAST-in-CUDA :

[0, 0, 03/18/2025]

[643, 2, 2025-03-18 17:41:59+00:00, 2019-09-02 15:10:41+00:00]

https://github.com/eurostat/statistics-coded :

[42, 16, 03/20/2025]

[527, 7, 2025-03-20 13:04:12+00:00, 2019-01-11 17:03:26+00:00]

https://github.com/tnagler/VineCopula :

[101, 34, 08/28/2026]

[562, 9, 2025-08-07 13:35:26+00:00, 2015-09-23 19:33:27+00:00]

https://github.com/deeptime-ml/deeptime :

[890, 89, 06/26/2026]

[1319, 17, 2025-04-14 02:31:18+00:00, 2015-04-11 10:31:18+00:00]

https://github.com/aiidateam/aiida-tutorials :

[25, 37, 10/02/2025]

[573, 31, 2025-10-02 11:28:57+00:00, 2018-05-28 16:53:57+00:00]

https://github.com/lepy/phuzzy :

[3, 2, 07/17/2026]

[267, 3, 2025-12-31 11:16:57+00:00, 2017-06-28 07:38:13+00:00]

https://github.com/dpiparo/vdt :

[61, 18, 04/03/2025]

[164, 6, 2025-05-03 22:18:58+00:00, 2012-08-23 07:16:32+00:00]